File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/SearchStore.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.service.store.database.AssetListDao
import com.gemwallet.android.data.service.store.database.SearchDao
import com.gemwallet.android.data.service.store.database.entities.DbSearch
import com.gemwallet.android.data.service.store.database.entities.toRecord
import com.gemwallet.android.serializer.decodeJson
import com.wallet.core.primitives.AssetList
import uniffi.gemstone.GemSearchStore

class GemstoneSearchStore(
    private val searchDao: SearchDao,
    private val assetListDao: AssetListDao,
) : GemSearchStore {
    override suspend fun setAssets(key: String, assetIds: List<String>) {
        if (assetIds.isEmpty()) {
            searchDao.deleteAssets(key)
        } else {
            searchDao.put(assetIds.mapIndexed { index, id -> DbSearch(query = key, assetId = id, priority = index) })
        }
    }

    override suspend fun setPerpetuals(key: String, perpetualIds: List<String>) {
        if (perpetualIds.isEmpty()) {
            searchDao.deletePerpetuals(key)
        } else {
            searchDao.put(perpetualIds.mapIndexed { index, id -> DbSearch(query = key, perpetualId = id, priority = index) })
        }
    }

    override suspend fun setLists(key: String, lists: List<String>) {
        val items = lists.map { it.decodeJson<AssetList>() }
        if (items.isEmpty()) {
            searchDao.deleteLists(key)
        } else {
            assetListDao.upsert(items.toRecord())
            searchDao.put(items.mapIndexed { index, list -> DbSearch(query = key, listId = list.id, priority = index) })
        }
    }
}

File: ios/Packages/GemstoneServices/Sources/Stores/SearchStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.AssetId
import typealias Gemstone.AssetList
import protocol Gemstone.GemSearchStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneSearchStore: GemSearchStore, @unchecked Sendable {
    private let store: SearchStore
    private let assetListStore: AssetListStore

    public init(store: SearchStore, assetListStore: AssetListStore) {
        self.store = store
        self.assetListStore = assetListStore
    }

    public func setAssets(key: String, assetIds: [Gemstone.AssetId]) async throws {
        try store.add(type: .asset, query: key, ids: assetIds)
    }

    public func setPerpetuals(key: String, perpetualIds: [String]) async throws {
        try store.add(type: .perpetual, query: key, ids: perpetualIds)
    }

    public func setLists(key: String, lists: [Gemstone.AssetList]) async throws {
        let lists = try lists.map { try Primitives.AssetList($0) }
        try assetListStore.upsert(lists)
        try store.add(type: .list, query: key, ids: lists.map(\.id))
    }
}
